stl/inc/ranges

```diff
@@ -4749,6 +4752,400 @@ namespace ranges {
+
+        template <bool _Const>
+        class _Sentinel;
+
+        template <class _Outer, class _Inner, class _Pattern, bool _Deref_is_glvalue>
+        struct _Category_base {};
+
+        template <forward_range _Outer, forward_range _Inner, class _Pattern>
+        struct _Category_base<_Outer, _Inner, _Pattern, true> {
+            using iterator_category = conditional_t<
+                !is_lvalue_reference_v<common_reference_t<range_reference_t<_Inner>, range_reference_t<_Pattern>>>,
+                input_iterator_tag,
+                conditional_t<common_range<_Inner> && common_range<_Pattern> //
+                                  && derived_from<_Iter_cat_t<iterator_t<_Outer>>, bidirectional_iterator_tag> //
+                                  && derived_from<_Iter_cat_t<iterator_t<_Inner>>, bidirectional_iterator_tag> //
+                                  && derived_from<_Iter_cat_t<iterator_t<_Pattern>>, bidirectional_iterator_tag>,
+                    bidirectional_iterator_tag,
+                    conditional_t<derived_from<_Iter_cat_t<iterator_t<_Outer>>, forward_iterator_tag> //
+                                      && derived_from<_Iter_cat_t<iterator_t<_Inner>>, forward_iterator_tag> //
+                                      && derived_from<_Iter_cat_t<iterator_t<_Pattern>>, forward_iterator_tag>,
+                        forward_iterator_tag, input_iterator_tag>>>;
+        };
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<_Maybe_const<_Const, _Vw>, _InnerRng<_Const>,
+                              _Maybe_const<_Const, _Pat>, is_reference_v<_InnerRng<_Const>>> {
+        private:
+            friend join_with_view;
+
+            template <bool>
+            friend class _Iterator;
+
+            template <bool>
+            friend class _Sentinel;
+
+            using _Parent_t    = _Maybe_const<_Const, join_with_view>;
+            using _Base        = _Maybe_const<_Const, _Vw>;
+            using _PatternBase = _Maybe_const<_Const, _Pat>;
+
+            using _OuterIter   = iterator_t<_Base>;
+            using _InnerIter   = iterator_t<_InnerRng<_Const>>;
+            using _PatternIter = iterator_t<_PatternBase>;
+
+            // True if and only if the expression *i, where i is an iterator from the outer range, is a glvalue:
+            static constexpr bool _Deref_is_glvalue = is_reference_v<_InnerRng<_Const>>;
+
+            _Parent_t* _Parent{};
+            /* [[no_unique_address]] */ _OuterIter _Outer_it{};
+            variant<_PatternIter, _InnerIter> _Inner_it{};
+
+            constexpr _Iterator(_Parent_t& _Parent_, iterator_t<_Base> _Outer_)
+                : _Parent{_STD addressof(_Parent_)}, _Outer_it{_STD move(_Outer_)} {
+                if (_Outer_it != _RANGES end(_Parent->_Range)) {
+                    auto&& _Inner = _Update_inner();
+                    _Inner_it.template emplace<1>(_RANGES begin(_Inner));

```

Comment:
I think, we should be able to reuse `_Variantish_state` here and throughout. We maybe need to rename it somewhat